In [3]:
import base64
from IPython.display import Image, display, HTML
import random
import time
from momaapi import MOMA, AnnVisualizer

dir_moma = "/Users/akashaj/Work/Datasets/MOMA-LRG"
# dir_moma = "/local-scratch/localhome/aabdujyo/Downloads/MOMA-LRG"
moma = MOMA(dir_moma)

visualizer = AnnVisualizer(moma)

In [ ]:
ids_sact = moma.get_ids_sact(split='test')

In [ ]:
print(len(ids_sact))

In [ ]:
id_sact = "02341"

ann_sact = moma.get_anns_sact(ids_sact=[id_sact])[0]
ids_hoi = moma.get_ids_hoi(ids_sact=[id_sact])
anns_hoi = moma.get_anns_hoi(ids_hoi=ids_hoi)

print(ann_sact)
print(ann_sact.length)
print(len(ids_hoi))
print(ids_hoi)
print(anns_hoi)
print(len(anns_hoi))


In [8]:
print(anns_hoi[1].__dir__())
print(anns_hoi[1].actors)
print(anns_hoi[1].objects)

for actor in anns_hoi[1].actors:
    print(actor.id)
    print(actor.kind)
    print(actor.cname)
    print(actor.cid)
    print(actor.bbox)
    
for obj in anns_hoi[1].objects:
    print(obj.id)
    print(obj.kind)
    print(obj.cname)
    print(obj.cid)
    print(obj.bbox)

['id', 'time', 'actors', 'objects', 'ias', 'tas', 'atts', 'rels', '__module__', '__doc__', '__init__', 'ids_actor', 'ids_object', '__repr__', '__dict__', '__weakref__', '__hash__', '__str__', '__getattribute__', '__setattr__', '__delattr__', '__lt__', '__le__', '__eq__', '__ne__', '__gt__', '__ge__', '__new__', '__reduce_ex__', '__reduce__', '__subclasshook__', '__init_subclass__', '__format__', '__sizeof__', '__dir__', '__class__']
[Actor(id=A, cname=basketball player), Actor(id=B, cname=basketball player), Actor(id=C, cname=basketball player), Actor(id=D, cname=basketball player)]
[Object(id=1, cname=basketball)]
A
actor
basketball player
3
BBox(x=154, y=82, w=186, h=398)
B
actor
basketball player
3
BBox(x=407, y=18, w=208, h=458)
C
actor
basketball player
3
BBox(x=37, y=138, w=176, h=340)
D
actor
basketball player
3
BBox(x=323, y=216, w=101, h=262)
1
object
basketball
13
BBox(x=367, y=297, w=34, h=37)


### Count frames, total number of boxes (people + objects) for each SAct and print stats

In [9]:
num_frames, num_bboxes, num_classes = [], [], []
ids_and_boxNum = []

start_time = time.time()

for id_sact in ids_sact:
    ann_sact = moma.get_anns_sact(ids_sact=[id_sact])[0]
    ids_hoi = moma.get_ids_hoi(ids_sact=[id_sact])
    anns_hoi = moma.get_anns_hoi(ids_hoi=ids_hoi)
    
    num_frames.append(len(ids_hoi))
    boxes, classes = 0, 0
    for ann_hoi in anns_hoi:
        boxes += len(ann_hoi.actors) + len(ann_hoi.objects)

    num_bboxes.append(boxes)
    ids_and_boxNum.append((id_sact, boxes))
    # num_classes.append(len(ids_hoi))

end_time = time.time()
execution_time = end_time - start_time

print(f"Execution time: {execution_time} seconds")

print("num of frames range:", min(num_frames), max(num_frames))
print("boxes range: ", min(num_bboxes), max(num_bboxes))

num_frames.sort()
num_bboxes.sort()

# print(len(num_frames),  num_bboxes[:300], num_bboxes[-500:])
n = [n for n in num_bboxes if n >= 10 and n <= 250]
# print(len(n),  n[:300], n[-500:])

ids_and_boxNum_selected = [(id_sact, boxes) for (id_sact, boxes) in ids_and_boxNum if boxes >= 10]

print(len(ids_and_boxNum), len(ids_and_boxNum_selected))

Execution time: 13.984117984771729 seconds
num of frames range: 4 32
boxes range:  0 963
3186 3120


In [10]:
ids_hoi = moma.get_ids_hoi(ids_sact=ids_sact)
anns_hoi = moma.get_anns_hoi(ids_hoi)
print(len(anns_hoi))

30323


### Class names and ids

In [9]:
print(anns_hoi[0])
print(ann_sact)
print(ann_sact.__dir__())
print(ann_sact.cname)
print(ann_sact.start)
print(ann_sact.end)
print(ann_sact.length)


print(ann_sact.id)
print(ann_sact.ids_actor)
print(ann_sact.ids_object)
for aact_actor in ann_sact.aacts_actor:
    print(aact_actor.id_entity)
    print(aact_actor.cname_entity)

for aact_object in ann_sact.aacts_object:
    print(aact_object.id_entity)
    print(aact_object.cname_entity)


NameError: name 'anns_hoi' is not defined

In [11]:
cids = moma.get_cids('sact', 0, 'train')
print(len(cids))
# print(cids)
cnames = moma.get_cnames(cids_sact=cids)
# print(len(cnames), cnames)

print(cnames.index(ann_sact.cname))

91


In [16]:
cids = moma.get_cids('actor', 0, 'train')
print(len(cids))
cids = moma.get_cids('actor', 0, 'test')
print(len(cids))
cids = moma.get_cids('actor', 0, 'val')
print(len(cids))

cnames = moma.get_cnames(cids_actor=cids)
print(cnames)

cids = moma.get_cids('object', 0, 'train')
print(len(cids))
cids = moma.get_cids('object', 0, 'test')
print(len(cids))
cids = moma.get_cids('object', 0, 'val')
print(len(cids))

cnames = moma.get_cnames(cids_object=cids)
print(cnames)

26
26
26
['award winner', 'babysitter', 'barber', 'basketball player', 'cashier', 'crowd', 'customer', 'driver', 'firefighter', 'frisbee player', 'health worker', 'match official', 'patient', 'pit crew member', 'police officer', 'presenter', 'receptionist', 'salon worker', 'soccer player', 'student', 'table tennis player', 'teacher', 'unclassified adult', 'unclassified athlete', 'unclassified child', 'waiter']
227
227
227
['alcohol', 'alcohol detector', 'alcohol pad', 'animal', 'arena', 'awards podium', 'baby basket', 'baby bottle', 'backpack', 'ball', 'barrel', 'basin', 'basket', 'basketball', 'basketball hoop', 'basketball stand', 'bath bucket', 'bed', 'belt', 'bench', 'bicycle', 'big car', 'blanket', 'board', 'book', 'bottle', 'bottle', 'bottle cap', 'bowl', 'box', 'bread', 'breathalyzer', 'brush', 'bucket', 'cabinet', 'camera', 'cap', 'car', 'car seat', 'card', 'card recorder', 'carpet', 'cash', 'cash register', 'cell phone', 'chair', "child's bicycle", "child's chair", 'chopsticks

### Check video size

In [5]:
# From box annotations

max_width, max_height = 0, 0

for id_sact in ids_sact:
    ann_sact = moma.get_anns_sact(ids_sact=[id_sact])[0]
    ids_hoi = moma.get_ids_hoi(ids_sact=[id_sact])
    anns_hoi = moma.get_anns_hoi(ids_hoi=ids_hoi)
    
    for ann_hoi in anns_hoi:        
        for actor in ann_hoi.actors:
            max_width = max(max_width, actor.bbox.x + actor.bbox.width)
            max_height = max(max_height, actor.bbox.y + actor.bbox.height)
            # print(actor)
            
        for obj in ann_hoi.objects:
            max_width = max(max_width, obj.bbox.x + obj.bbox.width)
            max_height = max(max_height, obj.bbox.y + obj.bbox.height)
                        
print(max_height, max_width)

720 1280


In [6]:
# From annotation file, gives the original video sizes

# Load data file here

import json
import os 

with open(os.path.join(dir_moma,'anns','anns.json'), 'r') as json_file:
    data = json.load(json_file)

print(len(data))

1412


In [7]:
# Explore data json
print(data[0].keys())
print(data[0]['activity']['sub_activities'][0].keys())

for item in data[:10]: 
    print(item['width'], item['height'], item['width']/item['height'])

print(min(data, key=lambda x: x['height'])['height']) 
print(max(data, key=lambda x: x['height'])['height']) 

print(min(data, key=lambda x: x['width'])['width']) 
print(max(data, key=lambda x: x['width'])['width']) 

min_aspect_ratio = min(data, key=lambda x: x['width']/x['height'])
max_aspect_ratio = max(data, key=lambda x: x['width']/x['height'])
print(min_aspect_ratio['width']/min_aspect_ratio['height'])
print(max_aspect_ratio['width']/max_aspect_ratio['height'])    

dict_keys(['file_name', 'num_frames', 'width', 'height', 'duration', 'activity'])
dict_keys(['id', 'class_name', 'start_time', 'end_time', 'higher_order_interactions'])
1280 720 1.7777777777777777
1280 720 1.7777777777777777
640 480 1.3333333333333333
1280 720 1.7777777777777777
854 480 1.7791666666666666
654 480 1.3625
1280 720 1.7777777777777777
1280 720 1.7777777777777777
1280 720 1.7777777777777777
1280 720 1.7777777777777777
360
720
462
1692
1.0
2.35


### Map between sact id and act id (to find video frame height)

In [8]:
from momaapi.lookup import Lookup
from momaapi.taxonomy import Taxonomy

taxonomy = Taxonomy(dir_moma)
lookup = Lookup(dir_moma, taxonomy, reset_cache=False)

In [11]:
id_sact = "02341"
id_act = lookup.map_id(id_sact=id_sact, kind='id_act')
# print(id_act)

act = list(filter(lambda x: x['file_name'].split('.')[0]==id_act, data))
print(id_act, len(act), act[0]['file_name'], act[0]['width'], act[0]['height'], 360/act[0]['height'])
print(act[0].keys())

Z_H_cQM2YRA 1 Z_H_cQM2YRA.mp4 640 480 0.75
dict_keys(['file_name', 'num_frames', 'width', 'height', 'duration', 'activity'])


In [79]:
# print(data[0].keys())
# print(data[0]['file_name'].split('.')[0])

# act = list(filter(lambda x: x['file_name'].split('.')[0]==id_act, data))
# print(len(act), act)

### Get data for dataloader

In [2]:
# From box annotations
import torch

class_ids = moma.get_cids('sact', 0, 'train') # total 91 classes
sact_class_names = moma.get_cnames(cids_sact=class_ids)
actor_class_ids = moma.get_cids('actor', 0, 'train') # total 91 classes
actor_class_names = moma.get_cnames(cids_actor=actor_class_ids)
object_class_ids = moma.get_cids('object', 0, 'train') # total 91 classes
object_class_names = moma.get_cnames(cids_object=object_class_ids)


bbox_set = torch.zeros(50, 289)
bbox_mask = torch.zeros(50, 1)

for id_sact in ids_sact[:1]:
    ann_sact = moma.get_anns_sact(ids_sact=[id_sact])[0]
    ids_hoi = moma.get_ids_hoi(ids_sact=[id_sact])
    anns_hoi = moma.get_anns_hoi(ids_hoi=ids_hoi)
    
    count_bbox = 0
    for idx, ann_hoi in enumerate(anns_hoi):  
        for actor in ann_hoi.actors:
            # print(actor.cname)
            bbox_set[count_bbox, idx] = 1 # frame number      
            actor_label_idx = actor_class_names.index(actor.cname) # total 26 classes
            bbox_set[count_bbox, actor_label_idx+36] = 1 # 32 + 4 = 36
            bbox_set[count_bbox, 32] = actor.bbox.x
            bbox_set[count_bbox, 33] = actor.bbox.y
            bbox_set[count_bbox, 34] = actor.bbox.width
            bbox_set[count_bbox, 35] = actor.bbox.height

            count_bbox += 1

        for obj in ann_hoi.objects:
            # print(obj.cname)
            bbox_set[count_bbox, idx] = 1 # frame number      
            obj_label_idx = object_class_names.index(obj.cname) # total 227 classes
            bbox_set[count_bbox, obj_label_idx+62] = 1 # 32 + 4 + 26 = 62
            bbox_set[count_bbox, 32] = obj.bbox.x
            bbox_set[count_bbox, 33] = obj.bbox.y
            bbox_set[count_bbox, 34] = obj.bbox.width
            bbox_set[count_bbox, 35] = obj.bbox.height

            count_bbox += 1
                        
    # for i in range(40):
    #     print(bbox_set[i])
    

NameError: name 'moma' is not defined

In [1]:
a = [1,2,3,4]
print(a.index(1))

0


### Saving csv files for video label annotations (for VideoMAE)

In [25]:
ids_sact = moma.get_ids_sact(split='test')
# ids_act = moma.get_ids_act(split='test')
ann_sact = moma.get_anns_sact(ids_sact=ids_sact)
# print(len(ann_sact))

# print(ann_sact[:20])

In [43]:
class_ids = moma.get_cids('sact', 0, split)
cnames = moma.get_cnames(cids_sact=class_ids)
cid_dict = dict(zip(cnames, class_ids))
# print(cid_dict)

{'one person is on their knees while proposing': 0, 'one person is putting a ring on their partner': 1, 'players are trying to catch a frisbee': 2, 'the adult is chasing after the child on the bike': 3, 'the adult is comforting the child': 4, 'the adult is explaining to the child how to ride a bike': 5, 'the adult is feeding the child': 6, 'the adult is holding the child on the bike': 7, 'the adult is playing with the child': 8, 'the adult is pushing the stroller': 9, "the adult is wrapping the child or changing the child's diaper or clothes": 10, 'the audience and host are cheering the award recipient on stage': 11, 'the award winner is delivering a speech on stage': 12, 'the barber is applying shampoo': 13, 'the barber is combing hair': 14, 'the barber is cutting hair with a razor': 15, 'the barber is cutting hair with scissors': 16, 'the barber is drying hair': 17, "the barber is massaging the customer's face or applying products": 18, 'the barber is shaving a beard with a razor': 1

#### Sub-activities

In [47]:
import csv

In [56]:
root_dir = "/home/aabdujyo/scratch/MOMA-LRG/sub_activity/"

for split in ['train', 'val', 'test']:
    ids_sact = moma.get_ids_sact(split=split)
    # ids_act = moma.get_ids_act(split='test')

    # Make a dictionary to get class ids from names
    class_ids = moma.get_cids('sact', 0, split)
    cnames = moma.get_cnames(cids_sact=class_ids)
    cid_dict = dict(zip(cnames, class_ids))
    
    with open(split+'_sact.csv', 'w') as csvfile:
        writer = csv.writer(csvfile, delimiter=' ')

        # For each item, find path to video file and get class id
        for id_sact in ids_sact:
            sact_path = root_dir + moma.get_paths(ids_sact = [id_sact])[0].split("/")[-1]
            sact_cname = moma.get_anns_sact(ids_sact=[id_sact])[0].cname    
            sact_cid = cid_dict[sact_cname]

    #         print(sact_path)
    #         print(sact_cname)
    #         print(sact_cid)
            writer.writerow([sact_path, sact_cid])
        
#     print("\n")

#### Activities

In [57]:
root_dir = "/home/aabdujyo/scratch/MOMA-LRG/activity/"

for split in ['train', 'val', 'test']:
    ids_act = moma.get_ids_act(split=split)
    # ids_act = moma.get_ids_act(split='test')

    # Make a dictionary to get class ids from names
    class_ids = moma.get_cids('act', 0, split)
    cnames = moma.get_cnames(cids_act=class_ids)
    cid_dict = dict(zip(cnames, class_ids))
#     print(cid_dict)
    
    with open(split+'_act.csv', 'w') as csvfile:
        writer = csv.writer(csvfile, delimiter=' ')

        # For each item, find path to video file and get class id
        for id_act in ids_act:
            act_path = root_dir + moma.get_paths(ids_act = [id_act])[0].split("/")[-1]
            act_cname = moma.get_anns_act(ids_act=[id_act])[0].cname    
            act_cid = cid_dict[act_cname]

#             print(act_path)
#             print(act_cname)
#             print(act_cid)
            writer.writerow([act_path, act_cid])
        
#     print("\n")

In [ ]:
import torch

In [7]:
a = torch.ones(1)*24
print(a)

tensor([24.])


## For V2 - based on Box Detection 

###### (Detection -> box encoder training -> boxPT -> FT)

#### Exploring saved data tensors

In [2]:
from torch.utils.data import DataLoader,TensorDataset

In [19]:
def get_dataset(split):
    data = torch.load(f'../data/{split}.pt')
    print(data.keys())
    print(data['bbox_sets'].size(), data['bbox_masks'].size(), data['sact_label_indices'].size())
    dataset = TensorDataset(data['bbox_sets'], 
                            data['bbox_masks'], 
                            data['sact_label_indices'])
    
    print(f"Loaded {split} data with {data['bbox_sets'].size(0)} items.")

    return dataset

In [50]:
train_data = DataLoader(get_dataset('train'), batch_size=8)
val_data = DataLoader(get_dataset('val'), batch_size=8)

# This dataset only has samples with >= 10 boxes annotated per sub-activity

dict_keys(['bbox_sets', 'bbox_masks', 'sact_label_indices'])
torch.Size([9858, 1142, 289]) torch.Size([9858, 1142]) torch.Size([9858, 1])
Loaded train data with 9858 items.
dict_keys(['bbox_sets', 'bbox_masks', 'sact_label_indices'])
torch.Size([2609, 1142, 289]) torch.Size([2609, 1142]) torch.Size([2609, 1])
Loaded val data with 2609 items.


In [52]:
for item in train_data:
    print(len(item))
    print(item[0].size(), item[1].size(), item[2].size())
    print(item[1])
    break

3
torch.Size([8, 1142, 289]) torch.Size([8, 1142]) torch.Size([8, 1])
tensor([[False, False, False,  ...,  True,  True,  True],
        [False, False, False,  ...,  True,  True,  True],
        [False, False, False,  ...,  True,  True,  True],
        ...,
        [False, False, False,  ...,  True,  True,  True],
        [False, False, False,  ...,  True,  True,  True],
        [False, False, False,  ...,  True,  True,  True]])


In [53]:
valdata = torch.load(f'../data/val.pt')


In [54]:
valdata['bbox_masks']src_key_padding_mask

tensor([[False, False, False,  ...,  True,  True,  True],
        [False, False, False,  ...,  True,  True,  True],
        [False, False, False,  ...,  True,  True,  True],
        ...,
        [False, False, False,  ...,  True,  True,  True],
        [False, False, False,  ...,  True,  True,  True],
        [False, False, False,  ...,  True,  True,  True]])

#### Creating a dict of { sact_id : [frame_ids] } - No need, all files in interaction folder are used

In [68]:
# ids_sact_train = moma.get_ids_sact(split='train')
# ids_sact_val = moma.get_ids_sact(split='val')
# ids_sact_test = moma.get_ids_sact(split='test')
ids_sact = moma.get_ids_sact()
print(len(ids_sact))

15842


In [86]:
id_sact = ids_sact[1]
print(id_sact)

00001


In [71]:
ids_hoi = moma.get_ids_hoi(ids_sact=[id_sact])

print(ids_hoi)

fnames = [f'/Users/akashaj/Work/Datasets/MOMA-LRG/interaction/{x}.jpg' for x in ids_hoi]
print(fnames)

6
['00001_0000000000', '00001_0001000000', '00001_0002000000', '00001_0003000000', '00001_0004000000', '00001_0004966667']
['/Users/akashaj/Work/Datasets/MOMA-LRG/interaction/00001_0000000000.jpg', '/Users/akashaj/Work/Datasets/MOMA-LRG/interaction/00001_0001000000.jpg', '/Users/akashaj/Work/Datasets/MOMA-LRG/interaction/00001_0002000000.jpg', '/Users/akashaj/Work/Datasets/MOMA-LRG/interaction/00001_0003000000.jpg', '/Users/akashaj/Work/Datasets/MOMA-LRG/interaction/00001_0004000000.jpg', '/Users/akashaj/Work/Datasets/MOMA-LRG/interaction/00001_0004966667.jpg']


In [73]:
ids_hoi = moma.get_ids_hoi(ids_sact=ids_sact)

In [75]:
print(len(ids_hoi))

161265


#### Checking stats (num of bboxes..) of detections

In [6]:
import json
import glob

root = "/Users/akashaj/Google Drive/My Drive/bbox_jsons/"

In [7]:
bbox_dict = {}
l = []

fnames = glob.glob(root + "*.json")
for fname in fnames:
    with open(fname) as json_file:
        data = json.load(json_file)
        
    bbox_dict[list(data.keys())[0]] = data[list(data.keys())[0]]
    l.append(len(bbox_dict[list(data.keys())[0]]))

In [8]:
print(len(l), max(l), min(l))

print(len([x for x in l if x>=5]))

print(len([x for x in l if x==4]))

print(bbox_dict['00024'][0:10])

15842 1187 0
15820
9
[[0, 165.19808959960938, 46.197288513183594, 304.4316711425781, 324.51812744140625, 0], [0, 87.4731216430664, 1.9580860137939453, 299.6953125, 211.45265197753906, 0], [0, 532.0689086914062, 95.90848541259766, 637.8604736328125, 209.9666290283203, 1], [0, 141.41712951660156, 168.072265625, 377.06646728515625, 351.608642578125, 1], [1, 87.62371063232422, 4.684548377990723, 336.89410400390625, 216.97396850585938, 0], [1, 193.30072021484375, 49.127655029296875, 320.633544921875, 285.9326171875, 0], [1, 148.25799560546875, 162.2445831298828, 406.9059143066406, 355.2938537597656, 1], [1, 154.29812622070312, 202.52902221679688, 197.65879821777344, 245.18368530273438, 0], [2, 122.29473114013672, 0.5318847298622131, 339.38671875, 273.5350341796875, 0], [2, 182.0887451171875, 156.69015502929688, 452.8713684082031, 356.7589416503906, 1]]


### Saving data for Box Encoder Training (save for min5 boxes, and min10 boxes)

    d = {"bbox_sets": torch.cat(bbox_set_full, dim=0),
         "bbox_masks": torch.cat(bbox_mask_full, dim=0),
         "sact_label_indices": torch.cat(sact_label_indices_full, dim=0)} 
bbox_sets -> 116 dimension [frame_num(one-hot), x, y, width, height, class(one-hot)] \ 

MOMA anotation:\
(1142, 289) \
1142 = Max number of bounding boxes per subactivity \
289 = 32   + 4 +   (26 + 227) 

MOMA detections:\
(1187, 116) \
1187 = Max number of bounding boxes per subactivity \
289 = 32   + 4 +   80 

In [23]:
# Getting sact class id (91 classes)

class_ids = moma.get_cids('sact', 0, 'train') # total 91 classes
sact_class_names = moma.get_cnames(cids_sact=class_ids)

# for fname in fnames:
print(fnames[0])
ann_sact = moma.get_anns_sact(ids_sact=[fnames[0][-10:-5]])[0]
print(ann_sact)
print(ann_sact.cname)
print(sact_class_names.index(ann_sact.cname))

/Users/akashaj/Google Drive/My Drive/bbox_jsons/data_00307.json
SAct(id=00307, cname=the driver is speaking with the police officer, time=[11.0, end=20.0), length=10)
the driver is speaking with the police officer
34


In [185]:
# Checking number of frames and detected class ids, and sact class_ids

class_ids = []
frame_nos = []
sact_class_ids = []
for fname in fnames:
    # bbox_dict[fnames[0][-10:-5]] --> List of bounding boxes for sact 
    # [[frame_num, x,y,w,h, class id]]
    for item in bbox_dict[fname[-10:-5]]:
        class_ids.append(item[5])
        frame_nos.append(item[0])
        
    ann_sact = moma.get_anns_sact(ids_sact=[fname[-10:-5]])[0]
    sact_class_ids.append(sact_class_names.index(ann_sact.cname))

print(max(class_ids))
print(max(frame_nos))
print(max(sact_class_ids))

79
31
90


In [150]:
import torch

In [226]:
# Do for train, val, test, and train-val
# For each min 5 and min 10 boxes per sub-activity

for min_boxes in [5, 10]:
    for split in ['train', 'val', 'test', 'train_val']:
        bbox_set_full, bbox_mask_full, sact_label_indices_full = [], [], []

        if split == 'train_val':
            ids_sact = moma.get_ids_sact(split='train')
            ids_sact.extend(moma.get_ids_sact(split='val'))
        else:
            ids_sact = moma.get_ids_sact(split=split)

        for id_sact in ids_sact:
            if len(bbox_dict[id_sact]) < min_boxes:
                continue

            bbox_set = torch.zeros(1187, 116)
            bbox_mask = torch.ones(1187)
            bbox_mask[:len(bbox_dict[id_sact])] = 0

            # Bounding box info [frame_num, x, y, w, h, object_label], mask, and box label
            for i, item in enumerate(bbox_dict[id_sact]):    
                bbox_set[i, item[0]] = 1
                bbox_set[i, 32] = item[1]
                bbox_set[i, 33] = item[2]
                bbox_set[i, 34] = item[3]
                bbox_set[i, 35] = item[4]
                bbox_set[i, 36 + item[5]] = 1 # box label, 80 classes, 32 + 4
            
            bbox_set_full.append(bbox_set.unsqueeze(0))
            bbox_mask_full.append(bbox_mask.unsqueeze(0))

            # Sub-activity label
            ann_sact = moma.get_anns_sact(ids_sact=[id_sact])[0]
            sact_label_indices_full.append(torch.tensor([sact_class_names.index(ann_sact.cname)]).unsqueeze(0))

        
        # bbox_set_full = torch.cat(bbox_set_full, dim=0)
        # bbox_mask_full = torch.cat(bbox_mask_full, dim=0)
        # sact_label_indices_full = torch.cat(sact_label_indices_full, dim=0)
        # print(bbox_set_full.size(), bbox_mask_full.size(), sact_label_indices_full.size())
        # print(bbox_set_full[:,:,32:36], bbox_mask_full, sact_label_indices_full)

        d = {"bbox_sets": torch.cat(bbox_set_full, dim=0),
         "bbox_masks": torch.cat(bbox_mask_full, dim=0),
         "sact_label_indices": torch.cat(sact_label_indices_full, dim=0)}
        
#         torch.save(d, f'/Users/akashaj/Work/activity_moma/detection_data/min{min_boxes}boxes/BoxEncoderTraining/{split}.pt')
        print(split, min_boxes, len(sact_label_indices_full))
        


train 5 9983
val 5 2656
test 5 3181
train_val 5 12639
train 10 9766
val 10 2574
test 10 3084
train_val 10 12340


#### Saving data as dict with fname keys for Video Box PreTraining

d[video_fname] = [bbox_set, bbox_mask, sact_label_index] \
video_fname : file name without the full path \
sact_label_idx = torch.tensor([self.sact_class_names.index(ann_sact.cname)]) # total 91 classes 

In [14]:
# Viewing previous video box pt data
import torch

data_path = "/Users/akashaj/Work/activity_moma/data_VideoBoxPT/test_sact.pt"

old_data = torch.load(data_path)

In [21]:
print(old_data['00000.mp4'])

print(old_data['00000.mp4'][0].size(), old_data['00000.mp4'][1].size(), old_data['00000.mp4'][2].size())

[tensor([[1., 0., 0.,  ..., 0., 0., 0.],
        [1., 0., 0.,  ..., 0., 0., 0.],
        [1., 0., 0.,  ..., 0., 0., 0.],
        ...,
        [0., 0., 0.,  ..., 0., 0., 0.],
        [0., 0., 0.,  ..., 0., 0., 0.],
        [0., 0., 0.,  ..., 0., 0., 0.]]), tensor([False, False, False,  ...,  True,  True,  True]), tensor([7])]
torch.Size([1142, 289]) torch.Size([1142]) torch.Size([1])


In [32]:
# Do for train, val, test, and train-val
# For min 10 boxes per sub-activity

for min_boxes in [10]:
    for split in ['train', 'val', 'test', 'train_val']:
        box_data_dict = {}
        if split == 'train_val':
            ids_sact = moma.get_ids_sact(split='train')
            ids_sact.extend(moma.get_ids_sact(split='val'))
        else:
            ids_sact = moma.get_ids_sact(split=split)

        for id_sact in ids_sact:
            if len(bbox_dict[id_sact]) < min_boxes:
                continue

            bbox_set = torch.zeros(1187, 116)
            bbox_mask = torch.ones(1187)
            bbox_mask[:len(bbox_dict[id_sact])] = 0

            # Bounding box info [frame_num, x, y, w, h, object_label], mask, and box label
            for i, item in enumerate(bbox_dict[id_sact]):    
                bbox_set[i, item[0]] = 1
                bbox_set[i, 32] = item[1]
                bbox_set[i, 33] = item[2]
                bbox_set[i, 34] = item[3]
                bbox_set[i, 35] = item[4]
                bbox_set[i, 36 + item[5]] = 1 # box label, 80 classes, 32 + 4
            
            # Sub-activity label
            ann_sact = moma.get_anns_sact(ids_sact=[id_sact])[0]
            
            box_data_dict[id_sact+'.mp4'] = [bbox_set, bbox_mask.to(torch.bool), torch.tensor([sact_class_names.index(ann_sact.cname)])]
        
        # bbox_set_full = torch.cat(bbox_set_full, dim=0)
        # bbox_mask_full = torch.cat(bbox_mask_full, dim=0)
        # sact_label_indices_full = torch.cat(sact_label_indices_full, dim=0)
        # print(bbox_set_full.size(), bbox_mask_full.size(), sact_label_indices_full.size())
        # print(bbox_set_full[:,:,32:36], bbox_mask_full, sact_label_indices_full)
        
        torch.save(box_data_dict, f'/Users/akashaj/Work/activity_moma/detection_data/min{min_boxes}boxes/VideoBoxPT/{split}.pt')        

In [33]:
# Viewing new detected box pt data

new_data_path = "/Users/akashaj/Work/activity_moma/detection_data/min10boxes/VideoBoxPT/test.pt"

new_data = torch.load(new_data_path)

In [34]:
print(old_data['00000.mp4'])

print(old_data['00000.mp4'][0].size(), old_data['00000.mp4'][1].size(), old_data['00000.mp4'][2].size())

print(new_data['00000.mp4'])

print(new_data['00000.mp4'][0].size(), new_data['00000.mp4'][1].size(), new_data['00000.mp4'][2].size())

[tensor([[1., 0., 0.,  ..., 0., 0., 0.],
        [1., 0., 0.,  ..., 0., 0., 0.],
        [1., 0., 0.,  ..., 0., 0., 0.],
        ...,
        [0., 0., 0.,  ..., 0., 0., 0.],
        [0., 0., 0.,  ..., 0., 0., 0.],
        [0., 0., 0.,  ..., 0., 0., 0.]]), tensor([False, False, False,  ...,  True,  True,  True]), tensor([7])]
torch.Size([1142, 289]) torch.Size([1142]) torch.Size([1])
[tensor([[1., 0., 0.,  ..., 0., 0., 0.],
        [1., 0., 0.,  ..., 0., 0., 0.],
        [1., 0., 0.,  ..., 0., 0., 0.],
        ...,
        [0., 0., 0.,  ..., 0., 0., 0.],
        [0., 0., 0.,  ..., 0., 0., 0.],
        [0., 0., 0.,  ..., 0., 0., 0.]]), tensor([False, False, False,  ...,  True,  True,  True]), tensor([7])]
torch.Size([1187, 116]) torch.Size([1187]) torch.Size([1])
